beacon-chain/db/pruner/pruner.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/db"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/db/iface"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/time/slots"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "db-pruner")
+
+type ServiceOption func(*Service)
+
+// WithRetentionPeriod allows the user to specify a different data retention period than the spec default.
+// The retention period is specified in epochs, and must be >= MIN_EPOCHS_FOR_BLOCK_REQUESTS.
+func WithRetentionPeriod(retentionEpochs primitives.Epoch) ServiceOption {
+	return func(s *Service) {
+		defaultRetentionEpochs := helpers.MinEpochsForBlockRequests() - 1
+		if retentionEpochs < defaultRetentionEpochs {
+			log.WithField("userEpochs", retentionEpochs).
+				WithField("minRequired", defaultRetentionEpochs).
+				Warn("Retention period too low, using minimum required value")
+		}
+
+		s.ps = pruneStartSlotFunc(retentionEpochs)
+	}
+}
+
+func WithSlotTicker(slotTicker slots.Ticker) ServiceOption {
+	return func(s *Service) {
+		s.slotTicker = slotTicker
+	}
+}
+
+type SyncChecker interface {
+	Synced() bool
+}
+
+type BackfillChecker interface {
+	IsComplete() bool
+}
+
+// Service defines a service that prunes beacon chain DB based on MIN_EPOCHS_FOR_BLOCK_REQUESTS.
+type Service struct {
+	ctx             context.Context
+	db              db.Database
+	ps              func(current primitives.Slot) primitives.Slot
+	prunedSlot      primitives.Slot
+	done            chan struct{}
+	slotTicker      slots.Ticker
+	syncChecker     SyncChecker
+	backfillChecker BackfillChecker
+}
+
+func New(ctx context.Context, db iface.Database, genesisTime time.Time, syncChecker SyncChecker, backfillChecker BackfillChecker, opts ...ServiceOption) (*Service, error) {
+	p := &Service{
+		ctx:             ctx,
+		db:              db,
+		ps:              pruneStartSlotFunc(helpers.MinEpochsForBlockRequests() - 1), // Default retention epochs is MIN_EPOCHS_FOR_BLOCK_REQUESTS - 1 from the current slot.
+		done:            make(chan struct{}),
+		slotTicker:      slots.NewSlotTicker(genesisTime, params.BeaconConfig().SecondsPerSlot),
+		syncChecker:     syncChecker,
+		backfillChecker: backfillChecker,
+	}
+
+	for _, o := range opts {
+		o(p)
+	}
+
+	return p, nil
+}
+
+func (p *Service) Start() {
+	log.Info("Starting Beacon DB pruner service")
+	go p.run()
+}
+
+func (p *Service) Stop() error {
+	log.Info("Stopping Beacon DB pruner service")
+	close(p.done)
+	return nil
+}
+
+func (p *Service) Status() error {
+	return nil
+}
+
+func (p *Service) run() {
+	defer p.slotTicker.Done()
+
+	for {
+		select {
+		case <-p.ctx.Done():
+			log.Debug("Stopping Beacon DB pruner service", "prunedSlot", p.prunedSlot)
+			return
+		case <-p.done:
+			log.Debug("Stopping Beacon DB pruner service", "prunedSlot", p.prunedSlot)
+			return
+		case slot := <-p.slotTicker.C():
```

Comment:
Nitpick: we're running a ticker every slot when we really only want to run exactly on slot 16 of the epoch, possibly emitting noisy logs on every other slot. Instead of using a slot ticker, what if the top of the loop was
```
for {
    untilMidEpoch := time.Duration(/*compute <time at next epoch + ((params.BeaconConfig().SlotsPerEpoch / 2) * seconds per epoch)> - time.Now */
    select {
        // handle ctx/done
        case <-time.After(untilMidEpoch):
            // pruning code
    }
}

Reply:
Tried this, but IMO slotticker is easily testable in unit tests, so i am going with that. Removed any noisy logs. Let me know your thoughts.